torch/optim/lr_scheduler.py

```diff
@@ -766,10 +766,12 @@ def __init__(self, schedulers):
                 )
         self._schedulers = list(schedulers)
         self.optimizer = schedulers[0].optimizer
+        self._last_lr = [group['lr'] for group in self._schedulers[-1].optimizer.param_groups]
 
     def step(self):
         for scheduler in self._schedulers:
             scheduler.step()
+        self._last_lr = [group['lr'] for group in self._schedulers[-1].optimizer.param_groups]
```

Comment:
Why do you always pick the last scheduler here?
I'm not even 100% convinced I understand the description of this class tbh.

Reply:
As far as I understand, `ChainedScheduler` does exactly the same thing as what would happen if we created multiple different schedulers and called their `step()` functions explicitly, as described under "How to adjust learning rate
" on this page: https://pytorch.org/docs/stable/optim.html?highlight=lr_scheduler

> Most learning rate schedulers can be called back-to-back (also referred to as chaining schedulers). The result is that each scheduler is applied one after the other on the learning rate obtained by the one preceding it.

So to me, it seems right for `ChainedScheduler.get_last_lr()` to return the learning rate from the last scheduler. However, maybe it would be more useful for it to return the learning rates from all the schedulers, combined in a list

Reply:
So it is very much possible that they override each other?
Thanks for the clarification!

Reply:
Right, the schedulers within a `ChainedScheduler` override each other, as long as they use the same optimizer. And they should always have the same optimizer, since we get an error if they have different optimizers:

```python
>>> import torch
>>> model = [torch.nn.parameter.Parameter(torch.randn(2, 2, requires_grad=True))]
>>> optimizer1 = torch.optim.SGD(model, 0.1)
>>> optimizer2 = torch.optim.SGD(model, 0.1)
>>> scheduler1 = torch.optim.lr_scheduler.ConstantLR(optimizer1, factor=0.1, total_iters=2)
>>> scheduler2 = torch.optim.lr_scheduler.ExponentialLR(optimizer2, gamma=0.9)
>>> scheduler = torch.optim.lr_scheduler.ChainedScheduler([scheduler1, scheduler2])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/work2/kurtamohler/development/pytorch-jit-utils-dtype/torch/optim/lr_scheduler.py", line 763, in __init__
    raise ValueError(
ValueError: ChainedScheduler expects all schedulers to belong to the same optimizer, but got schedulers at index 0 and 1 to be different
```